providers/apis/defi/raydium/provider.go

```diff
@@ -0,0 +1,252 @@
+package raydium
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"time"
+
+	binary "github.com/gagliardetto/binary"
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/programs/token"
+	"github.com/gagliardetto/solana-go/rpc"
+	"go.uber.org/zap"
+
+	"github.com/skip-mev/slinky/oracle/config"
+	oracletypes "github.com/skip-mev/slinky/oracle/types"
+	"github.com/skip-mev/slinky/pkg/math"
+	"github.com/skip-mev/slinky/providers/base/api/handlers"
+	providertypes "github.com/skip-mev/slinky/providers/types"
+	mmtypes "github.com/skip-mev/slinky/x/marketmap/types"
+)
+
+var _ handlers.APIPriceFetcher[mmtypes.Ticker, *big.Int] = (*APIPriceFetcher)(nil)
+
+// SolanaJSONRPCClient is the expected interface for a solana JSON-RPC client according
+// to the APIPriceFetcher.
+//
+//go:generate mockery --name SolanaJSONRPCClient --output ./mocks/ --case underscore
+type SolanaJSONRPCClient interface {
+	GetMultipleAccountsWithOpts(
+		ctx context.Context,
+		accounts []solana.PublicKey,
+		opts *rpc.GetMultipleAccountsOpts,
+	) (out *rpc.GetMultipleAccountsResult, err error)
+}
+
+// APIPriceFetcher is responsible for interacting with the solana API and querying information
+// about the price of a given currency pair.
+type APIPriceFetcher struct {
+	// market represents the ticker configurations for this provider.
+	market oracletypes.ProviderMarketMap
+
+	// config is the APIConfiguration for this provider
+	config config.APIConfig
+
+	// client is the solana JSON-RPC client used to query the API.
+	client SolanaJSONRPCClient
+
+	// metaDataPerTicker is a map of ticker.String() -> TickerMetadata
+	metaDataPerTicker map[string]TickerMetadata
+
+	// logger
+	logger *zap.Logger
+}
+
+// NewAPIPriceFetcher returns a new APIPriceFetcher. This method constructs the
+// default solana JSON-RPC client in accordance with the config's URL param.
+func NewAPIPriceFetcher(
+	market oracletypes.ProviderMarketMap,
+	config config.APIConfig,
+	logger *zap.Logger,
+) (*APIPriceFetcher, error) {
+	return NewAPIPriceFetcherWithClient(
+		market,
+		config,
+		rpc.New(config.URL),
+		logger,
+	)
+}
+
+// NewAPIPriceFetcherWithClient returns a new APIPriceFetcher. This method requires
+// that the given market + config are valid, otherwise a nil implementation + an error
+// will be returned.
+func NewAPIPriceFetcherWithClient(
+	market oracletypes.ProviderMarketMap,
+	config config.APIConfig,
+	client SolanaJSONRPCClient,
+	logger *zap.Logger,
+) (*APIPriceFetcher, error) {
+	if err := config.ValidateBasic(); err != nil {
+		return nil, err
+	}
+
+	if err := market.ValidateBasic(); err != nil {
```

Comment:
needs test case

Reply:
I'm testing this [here](https://github.com/skip-mev/slinky/pull/301/files#diff-e7c5b6ecb4619f7352f334ab7b4c7b5f046035f8a766924d5586dd80d329cd39R53) not sure why it doesn't show up in coverprof